test/e2e/major_upgrade_v3.go

```diff
@@ -97,12 +99,54 @@ func MajorUpgradeToV3(logger *log.Logger) error {
 				testnet.NoError("failed to get header", err)
 				if resp.Header.Version.App == v3.Version {
 					upgradeComplete = true
+					if upgradedHeight == 0 {
+						upgradedHeight = resp.Header.Height
+					}
 				}
 				logger.Printf("height %v", resp.Header.Height)
 				lastHeight = resp.Header.Height
 			}
 		}
 	}
 
+	// check if the timeouts are set correctly
+	rpcNode := testNet.Nodes()[0]
+	client, err := rpcNode.Client()
+	testnet.NoError("failed to get client", err)
+
+	startHeight := upgradeHeightV3 - 5
+	endHeight := upgradedHeight + 1
+	blockTimes := make([]time.Duration, 0, endHeight-startHeight)
+	var prevBlockTime time.Time
+
+	for h := startHeight; h < endHeight; h++ {
+		resp, err := client.Header(ctx, &h)
+		testnet.NoError("failed to get header", err)
+		blockTime := resp.Header.Time
+
+		if h == startHeight {
+			if resp.Header.Version.App != v2.Version {
+				return fmt.Errorf("expected start height %v was app version 2", startHeight)
+			}
+			prevBlockTime = blockTime
+			continue
+		}
+
+		blockDur := blockTime.Sub(prevBlockTime)
+		prevBlockTime = blockTime
+		blockTimes = append(blockTimes, blockDur)
+	}
+
+	startDur := blockTimes[0]
+	endDur := blockTimes[len(blockTimes)-1]
+
+	if startDur < time.Second*10 {
+		testnet.NoError("", fmt.Errorf("blocks for v2 are too short %v", len(blockTimes)))
+	}
+
+	if endDur > time.Second*7 {
+		testnet.NoError("", fmt.Errorf("blocks for v3 are too long %v", len(blockTimes)))
+	}
```

Comment:
So if I understand correctly, this just picks the duration of the block 5 before the signal height and 1 after the upgrade height and asserts the first is greater than 10 seconds and the last is less than 7 seconds.

I think it should check all the heights up to the actual upgrade height (i.e. during the delay period) to be sure that it still remains at the expected times

Reply:
refactored in 07c1fc98836d040ed48a9ae84e76969def2ab656